python/ql/src/experimental/semmle/python/frameworks/Xml.qll

```diff
@@ -8,298 +8,456 @@ private import semmle.python.dataflow.new.DataFlow
+    DataFlow::TypeBackTracker t, DataFlow::Node arg
   ) {
     t.start() and
-    exists(DataFlow::MethodCallNode featureCall |
-      featureCall = parser.getAMethodCall("setFeature") and
-      featureCall.getArg(0).getALocalSource() = feature.getAUse() and
-      featureCall.getArg(1).getALocalSource() = DataFlow::exprNode(any(True t_)) and
-      result = featureCall.getObject()
-    )
+    arg = any(SaxParserSetFeatureCall c).getStateArg() and
+    result = arg.getALocalSource()
     or
-    exists(DataFlow::TypeTracker t2 |
-      t = t2.smallstep(trackSaxFeature(t2, parser, feature), result)
+    exists(DataFlow::TypeBackTracker t2 |
+      result = saxParserSetFeatureStateArgBacktracker(t2, arg).backtrack(t2, t)
     )
   }
 
-  /** Gets a reference to a `parser` that has been set a `feature`. */
-  DataFlow::Node trackSaxFeature(DataFlow::CallCfgNode parser, API::Node feature) {
-    result = trackSaxFeature(DataFlow::TypeTracker::end(), parser, feature)
+  /** Gets a back-reference to the `setFeature` state argument `arg`. */
+  DataFlow::LocalSourceNode saxParserSetFeatureStateArgBacktracker(DataFlow::Node arg) {
+    result = saxParserSetFeatureStateArgBacktracker(DataFlow::TypeBackTracker::end(), arg)
   }
 
   /**
-   * Gets a call to `xml.sax.make_parser`.
+   * Gets a reference to a XML sax parser that has `feature_external_ges` turned on.class
    *
-   * Given the following example:
-   *
-   * ```py
-   * BadHandler = MainHandler()
-   * parser = xml.sax.make_parser()
-   * parser.setContentHandler(BadHandler)
-   * parser.setFeature(xml.sax.handler.feature_external_ges, False)
-   * parser.parse(StringIO(xml_content))
-   * parsed_xml = BadHandler._result
-   * ```
-   *
-   * * `this` would be `xml.sax.make_parser()`.
-   * * `getAnInput()`'s result would be `StringIO(xml_content)`.
-   * * `vulnerable(kind)`'s `kind` would be `Billion Laughs` and `Quadratic Blowup`.
+   * See https://docs.python.org/3/library/xml.sax.handler.html#xml.sax.handler.feature_external_ges
    */
-  private class XMLSaxParser extends DataFlow::CallCfgNode, XML::XMLParser::Range {
-    XMLSaxParser() {
-      this = API::moduleImport("xml").getMember("sax").getMember("make_parser").getACall()
-    }
+  private DataFlow::Node saxParserWithFeatureExternalGesTurnedOn(DataFlow::TypeTracker t) {
+    t.start() and
+    exists(SaxParserSetFeatureCall call |
+      call.getFeatureArg() =
+        API::moduleImport("xml")
+            .getMember("sax")
+            .getMember("handler")
+            .getMember("feature_external_ges")
+            .getAUse() and
+      saxParserSetFeatureStateArgBacktracker(call.getStateArg())
+          .asExpr()
+          .(BooleanLiteral)
+          .booleanValue() = true and
+      result = call.getObject()
+    )
+    or
+    exists(DataFlow::TypeTracker t2 |
+      t = t2.smallstep(saxParserWithFeatureExternalGesTurnedOn(t2), result)
+    ) and
```

Comment:
```suggestion
    (
      t.start() and
      exists(SaxParserSetFeatureCall call |
        call.getFeatureArg() =
          API::moduleImport("xml")
              .getMember("sax")
              .getMember("handler")
              .getMember("feature_external_ges")
              .getAUse() and
        saxParserSetFeatureStateArgBacktracker(call.getStateArg())
            .asExpr()
            .(BooleanLiteral)
            .booleanValue() = true and
        result = call.getObject()
      )
      or
      exists(DataFlow::TypeTracker t2 |
        t = t2.smallstep(saxParserWithFeatureExternalGesTurnedOn(t2), result)
      )
    ) and
```

I'm not sure if grouping this would help at all. If not, isn't the `and not` only affecting the smallstep recursion? 🤔

Reply:
> I'm not sure if grouping this would help at all. If not, isn't the and not only affecting the smallstep recursion? thinking

Yes, but that was also the intention. So you _could_ have this:

```py
parser = xml.sax.make_parser()

parser.setFeature(xml.sax.handler.feature_external_ges, True)
parser.parse(StringIO(x)) # XXE vuln on this line

parser.setFeature(xml.sax.handler.feature_external_ges, False)
parser.parse(StringIO(x)) # but not on this line
```

So the type-tracker will stop tracking the unsafe parser, if it is made safe again.

I think the only change from doing your version, would be in the example below. With the current version, we treat the parser as unsafe, and with your suggested version, we would not start tracking it since the value might be false. 

```py
parser = xml.sax.make_parser()

if cond:
    flag_value = True
else:
    flag_value = False

parser.setFeature(xml.sax.handler.feature_external_ges, flag_value)
parser.parse(StringIO(x)) # XXE vuln on this line???
```

This was not something I had thought about, but actually a subtle point to get right. I think we should treat this as vuln! That also means we handle it in the same way as the snippet below

```python
    parser = xml.sax.make_parser()
    if cond:
        parser.setFeature(xml.sax.handler.feature_external_ges, True)
    else:
        parser.setFeature(xml.sax.handler.feature_external_ges, False)
    parser.parse(StringIO(x)) # we track this as vuln to XXE
```

Reply:
These tests added in 1a9620a

Reply:
I double checked that my understanding was right, and scratched my head a bit initially when I could not produce the expected results. the culprit was splitting (where we keep multiple copies of the code in the DB, for different branches of an if-statement), and was resolved by not having both conditional assignment tests in the same function :face_exhaling: